### PR TITLE
[Float8Quant] Add rowwise scaling option to float8 dyanmic quant

### PR DESCRIPTION
**Pull request by drisspg**
Stacked PRs:
 * __->__#819

--- --- ---

[Float8Quant] Add rowwise scaling option to float8 dyanmic quant

# Summary
Adds an option to do PerRow scaling of the activation and weight, which can provide faster activation quantization and better numerics.

The api for calling this is:
```Python
import torch
import torch.nn as nn
from torchao.quantization.quant_api import quantize_, PerRow, float8_dynamic_activation_float8_weight

linear_module = nn.Linear(16, 32).to("cuda")

quantize_(linear_module, float8_dynamic_activation_float8_weight(granularity=PerRow())) 
```

### CODE REVIEW

File: torchao/quantization/quant_api.py [hunk 1]
```diff
@@ -550,6 +552,7 @@ def apply_float8wo_quant(weight):
 def float8_dynamic_activation_float8_weight(
     activation_dtype: torch.dtype = torch.float8_e4m3fn,
     weight_dtype: torch.dtype = torch.float8_e4m3fn,
+    granularity: Literal[PerTensor, RowWise] = PerTensor,
```
Comment (drisspg) [inline torchao/quantization/quant_api.py:555]: @jerryzh168 @vkuzo Thoughts on what the public API should be with this? I stole the granularity from the observers module. Note I could also do: PerAxis(0) as the only supported option

File: torchao/quantization/observer.py [hunk 2]
```diff
@@ -52,6 +52,17 @@ class PerAxis(GranularityType):
     """
     axis: int
 
+@dataclass(frozen=True)
+class RowWise(GranularityType):
```
Comment (vkuzo) [inline torchao/quantization/observer.py:56]: n00b question, why is GranularityType a set of classes instead of an enum?

if it's for storing granularity settings later, maybe `GranularityType` isn't the right name?  That name screams enum to me.

File: torchao/quantization/quant_api.py (hunk 1 above)
Comment (vkuzo) [reply torchao/quantization/quant_api.py:555]: would be good to have granularity for each tensor (for example, rowwise for input1 and tensorwise for input2).

File: torchao/quantization/observer.py (hunk 2 above)
Comment (drisspg) [reply torchao/quantization/observer.py:56]: I think because some subclasses want to store more information, I.e.  PerAxis stores which axis and a PerBlock would store block_size in each dim

Comment (vkuzo) [reply torchao/quantization/observer.py:56]: call it `GranularityConfig` then?

Comment (drisspg) [reply torchao/quantization/observer.py:56]: @jerryzh168 how does this sound, I am not sure if this API his heavily used else where or is it okay to just change the name?

Comment (jerryzh168) [reply torchao/quantization/observer.py]: GranularityConfig sounds reasonable I think, it's OK to change the name, it's only used in observer right now

Comment (jerryzh168) [reply torchao/quantization/observer.py]: or just `Granularity`

File: torchao/quantization/quant_api.py [hunk 3]
```diff
@@ -660,23 +698,46 @@ def float8_dynamic_activation_float8_weight(
     if mm_config is None:
         mm_config = Float8MMConfig(use_fast_accum=True)
 
-    #TODO we are hardcoding TensorWise scaling, will follow up PR for Tensorwise scaling
+    activation_granularity, weight_granularity = _validate_granularity(granularity)
+
+    def get_block_size(x: torch.Tensor, granularity: _fp8_granularities):
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:703]: why is this not added to the official one in https://github.com/pytorch/ao/blob/144445a7a8e988059555421caafa17e0c1678053/torchao/quantization/observer.py#L92

File: torchao/quantization/quant_api.py [hunk 4]
```diff
@@ -641,17 +643,53 @@ def apply_float8wo_quant(weight):
     return _get_linear_subclass_inserter(apply_float8wo_quant)
 
 
+_fp8_granularities = Literal[PerTensor, PerRow]
+
+
+# Validate and process granularity input
+def _validate_granularity(
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:650]: nit: maybe name should be `normalize_granularity`

File: torchao/quantization/quant_api.py [hunk 5]
```diff
@@ -65,6 +65,8 @@
 )
 from torchao.float8.inference import Float8MMConfig
 
+from torchao.quantization.observer import PerTensor, PerAxis, PerRow
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:68]: nit: PerAxis is not used

File: test/dtypes/test_affine_quantized_float.py [hunk 6]
```diff
@@ -68,33 +66,81 @@ class TestAffineQuantizedFloat8Compile(InductorTestCase):
...
+            ), f"Quantization error is too high got a SQNR of {error}"
+
+    def test_invalid_granularity(self):
+        with pytest.raises(ValueError, match="Invalid granularity specification"):
+            float8_dynamic_activation_float8_weight(granularity="invalid")
+
+    def test_mismatched_granularity(self):
+        with pytest.raises(
+            ValueError,
+            match="Different granularities for activation and weight are not supported",
```
Comment (jerryzh168) [inline test/dtypes/test_affine_quantized_float.py:120]: so will this be supported in the future?

**Review by jerryzh168 — APPROVED** [review]
LGTM!

File: torchao/quantization/observer.py (hunk 2 above)
Comment (drisspg) [reply torchao/quantization/observer.py]: Im going to do this in a follow up PR

File: test/dtypes/test_affine_quantized_float.py (hunk 6 above)
Comment (drisspg) [reply test/dtypes/test_affine_quantized_float.py:120]: Yup, going to add as a follow up, PerTensor + PerRowise should work if you broadcast the PerTensor to RowWise

File: torchao/quantization/quant_api.py (hunk 3 above)
Comment (drisspg) [reply torchao/quantization/quant_api.py:703]: Ohh great call!
